### PR TITLE
fix: #3095 adding detection of create conflicts

### PR DESCRIPTION
**Pull request by shawkins**
## Description
fix for #3095 - it requires a few tweaks to check for the existing items.

It does highlight an issue with raw attribute extraction - I've commented out the relevant test for now, but you could take this a step further and try to make that extraction work.

cc @rohanKanojia 

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [x] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [ ] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Review by oscerd — APPROVED** [review]

File: kubernetes-server-mock/src/test/java/io/fabric8/kubernetes/client/server/mock/KubernetesCrudAttributesExtractorTest.java [hunk 1]
```diff
@@ -326,6 +326,21 @@ public void statusHandling() {
...
+    kubernetesServer.before();
+    KubernetesClient kubernetesClient = kubernetesServer.getClient();
+    Pod pod = new PodBuilder().withNewMetadata()
+            .withName("name")
+            .withNamespace("test") // required until https://github.com/fabric8io/mockwebserver/pull/59
+            .endMetadata()
+            .withStatus(new PodStatusBuilder().withHostIP("x").build())
+            .build();
+    kubernetesClient.pods().create(pod);
+    assertThrows(KubernetesClientException.class, ()-> kubernetesClient.pods().create(pod));
```
Comment (rohanKanojia) [inline kubernetes-server-mock/src/test/java/io/fabric8/kubernetes/client/server/mock/KubernetesCrudAttributesExtractorTest.java:341]: nit, Please add an assertion for HTTP_CONFLICT code received in exception.
```
    MixedOperation<Pod, PodList, PodResource<Pod>> podOp = kubernetesClient.pods();
    podOp.create(pod);
    KubernetesClientException exception = assertThrows(KubernetesClientException.class, () -> podOp.create(pod));
    Assertions.assertEquals(HTTP_CONFLICT, exception.getCode());
```

File: kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/CustomResourceCrudTest.java [hunk 2]
```diff
@@ -108,9 +107,11 @@ void testCreateOrReplaceRaw() throws IOException {
...
     object.put("spec", "updated");
 
-    Map<String, Object> updated = raw.createOrReplace(object);
-    assertNotEquals(created, updated);
-    assertEquals(object, updated);
+    // with the mock detecting conflicts, this won't work because the
+    // appropriate attributes are extracted to perform the get after the 409
+    //Map<String, Object> updated = raw.createOrReplace(object);
+    //assertNotEquals(created, updated);
+    //assertEquals(object, updated);
```
Comment (rohanKanojia) [inline kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/CustomResourceCrudTest.java:114]: Does this mean createOrReplace will no longer work in crud mode?

Comment (shawkins) [reply kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/CustomResourceCrudTest.java:114]: More broadly it means that GET doesn't work in raw mode - the metadata is not extracted properly during create.

File: kubernetes-server-mock/src/test/java/io/fabric8/kubernetes/client/server/mock/KubernetesCrudAttributesExtractorTest.java (hunk 1 above)
Comment (shawkins) [reply kubernetes-server-mock/src/test/java/io/fabric8/kubernetes/client/server/mock/KubernetesCrudAttributesExtractorTest.java]: okay, I've made that update.

File: kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/CustomResourceCrudTest.java (hunk 2 above)
Comment (shawkins) [reply kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/CustomResourceCrudTest.java]: I've updated this pr with a fix to that problem and another change to not send update events when nothing changes.

**Review by manusa — APPROVED** [review]
LGTM, thx!

**Pull request merged**